### PR TITLE
Build: Upgrade to Gradle 7.x

### PR DESCRIPTION
**Pull request by nastra**
This PR was motivated by https://github.com/apache/iceberg/pull/2816, where @Rolly992 tried upgrading to Gradle 7.x but ran into a bunch of failures/issues.

Upgrading to Gradle 7.x was non-trivial and there were a bunch of things that were failing in the process. I think we should eventually upgrade to Gradle 7 as the longer we wait, the harder it will get in the future. However, we shouldn't merge this before releasing Iceberg 0.12.0 as that would impose an unnecessary risk.

It makes sense to review each commit individually. Noteworthy changes that were done are shown below:

* changes to `gradlew` are coming from the gradle wrapper upgrade
* The `compile` / `runtime` / `testRuntime` / `testCompile` configurations were removed and replaced with `implementation` / `runtimeOnly` / `testRuntimeOnly` / `testImplementation`. See also https://docs.gradle.org/current/userguide/upgrading_version_6.html#sec:configuration_removal for additional details. `compile` and `runtime` configurations did not support a fine grained scoping of dependencies previously and so it was possible that projects were compiling due to implicit dependencies. I had to change stuff to make those dependencies explicit now.
* `apply plugin: 'maven'` was removed due to https://docs.gradle.org/current/userguide/upgrading_version_6.html#removal_of_the_legacy_maven_plugin
* `java` plugin was replaced by `java-library` so that we can use the `api` configuration as noted in https://docs.gradle.org/current/userguide/java_library_plugin.html#java_library_plugin
* `com.palantir.baseline-scalastyle` uses https://github.com/ngbinh/gradle-scalastyle-plugin, which has been updated in April 2018 and isn't Gradle 7 ready, so I had to switch to https://github.com/alisiikh/gradle-scalastyle-plugin
* A newer checkstyle version is being used under the hood, therefore a few more places were failing and were updated
* `com.palantir.baseline-error-prone` uses a newer error-prone version (and does additional checks), so a few changes were necessary to fix `JavaTimeSystemDefaultTimeZone` / `ConsistentOverrides` errors
* After the upgrade I compared the Gradle-7-generated-jars with the Gradle-5-generated-jars. I noticed that the jars of the `xx-runtime.jar` is slightly **bigger** on this branch because `exclude '...'` isn't properly working due to https://github.com/johnrengelman/shadow/issues/505. The last commit (4e33561fd7c256d7c42e06d93d7f7e64fd5d6877) in the PR fixes this issue.

/cc @rymurr / @snazy / @rdblue

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 1]
```diff
@@ -454,6 +454,7 @@ public TableMetadata refresh() {
     }
 
     @Override
+    @SuppressWarnings("ConsistentOverrides")
     public void commit(TableMetadata underlyingBase, TableMetadata metadata) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:458]: renaming `underlyingBase` to `base` here does conflict and shadows the `base` variable that's defined in the class itself

File: build.gradle [hunk 2]
```diff
@@ -656,27 +707,37 @@ if (jdkVersion == '8') {
         exclude group: 'org.pentaho' // missing dependency
         exclude group: 'org.slf4j', module: 'slf4j-log4j12'
       }
+
+      compileOnly("org.apache.orc:orc-core::nohive") {
```
Comment (snazy) [inline build.gradle:711]: This kind of dependency w/ excludes is repeated. I think, it would be cleaner to add a new configuration named something like `orc-core-nohive` so so and add that configuration as a dependency instead of the repeated lines. WDYT?

**Review by snazy — COMMENTED** [review]
From a brief look, the change looks reasonable to me.

Can you compare the generated pom's from the current build-files and see whether there are noteworthy differences to the pom's generated with this change? And whether the generated shadow-jar(s) still have the same content and pom's? Issues in there can be tricky to investigate later.

**Comment by nastra** [comment]
> Can you compare the generated pom's from the current build-files and see whether there are noteworthy differences to the pom's generated with this change? And whether the generated shadow-jar(s) still have the same content and pom's? Issues in there can be tricky to investigate later.

Yes I did compare those already and I noticed that the jars of the `xx-runtime.jar` is slightly **bigger** on this branch because `exclude '...'` isn't properly working due to https://github.com/johnrengelman/shadow/issues/505

File: api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java [hunk 3]
```diff
@@ -429,7 +429,7 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
     }
 
     private boolean canContainNulls(Integer id) {
-      return nullCounts == null || (nullCounts.containsKey(id) && nullCounts.get(id) > 0);
+      return nullCounts == null || nullCounts.containsKey(id) && nullCounts.get(id) > 0;
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java:432]: I think it is much better to have the parentheses so that the logic is clear to a reader, even if it is identical to Java's precedence. Can you update the style config and revert these changes?

Comment (nastra) [reply api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java]: @rdblue pushed a new version where the `UnnecessaryParentheses` checkstyle module was removed

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetDefinitionLevelReader.java [hunk 4]
```diff
@@ -305,7 +305,7 @@ public void nextDictEncodedBatch(
     protected abstract void nextVal(
         FieldVector vector, int idx, ValuesAsBytesReader valuesReader, int typeWidth, byte[] byteArray);
     protected abstract void nextDictEncodedVal(
-        FieldVector vector, int startOffset, VectorizedDictionaryEncodedParquetValuesReader reader,
+        FieldVector vector, int idx, VectorizedDictionaryEncodedParquetValuesReader reader,
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetDefinitionLevelReader.java:308]: Do you know why this is needed now? I thought we already caught variable names that shadow field names.

File: build.gradle [hunk 5]
```diff
@@ -174,7 +175,7 @@ project(':iceberg-bundled-guava') {
 
   shadowJar {
     classifier null
-    configurations = [project.configurations.compileOnly]
+    configurations = [project.configurations.compileClasspath]
```
Comment (rdblue) [inline build.gradle:178]: Is this correct? We include `compileOnly` to avoid including `compile` or `implementation` dependencies that are expected in the runtime classpath.

File: build.gradle [hunk 6]
```diff
@@ -247,13 +250,30 @@ project(':iceberg-data') {
       exclude group: 'org.slf4j', module: 'slf4j-log4j12'
     }
 
-    testCompile("org.apache.hadoop:hadoop-client") {
+    implementation("org.apache.orc:orc-core::nohive") {
```
Comment (rdblue) [inline build.gradle:253]: These are no longer pulled in via iceberg-orc and iceberg-core? What about Parquet? Do we need an implementation dependency on Parquet as well?

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetDefinitionLevelReader.java (hunk 4 above)
Comment (nastra) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetDefinitionLevelReader.java:308]: this is due to `ConsistentOverrides` from the additional ErrorProne checks in https://github.com/palantir/gradle-baseline#baseline-error-prone-checks

File: build.gradle (hunk 5 above)
Comment (nastra) [reply build.gradle:178]: yes that's correct. I made sure that the generated Jars are identical

File: build.gradle (hunk 6 above)
Comment (nastra) [reply build.gradle:253]: > These are no longer pulled in via iceberg-orc and iceberg-core? 

correct, because `implementation` allows now more fine-grained control over dependencies being pulled in, which wasn't the case previously with `compile` (you got more dependencies than you were asking for). Now one could probably achieve the same thing by pulling in certain dependencies via `api`, but that means you're making those dependencies explicitly part of your API. The Gradle docs also state that one should generally prefer `implementation` over `api`, meaning it is better to explicitly state all the dependencies that are really required, so that you don't get surprises with non-expected dependencies and such.

> Do we need an implementation dependency on Parquet as well?
already added it in https://github.com/nastra/iceberg/blob/6c308fd66a9f33ce31b5abf8de61dd8269fda4b5/build.gradle#L261

Comment (nastra) [reply build.gradle:253]: or did you mean the one defined in https://github.com/nastra/iceberg/blob/6c308fd66a9f33ce31b5abf8de61dd8269fda4b5/build.gradle#L245?

Comment (rdblue) [reply build.gradle:253]: This adds implementation dependencies for org.apache.avro:avro and org.apache.orc:orc-core. Why isn't there one for org.apache.parquet?

Can you be more specific about what you're saying with implementation vs compile? When are transitive dependencies included and when are they not?

Comment (nastra) [reply build.gradle:253]: > This adds implementation dependencies for org.apache.avro:avro and org.apache.orc:orc-core. Why isn't there one for org.apache.parquet?

Judging by the 3 parquet imports that are being used in `:iceberg-data`, we have:

```
import org.apache.parquet.hadoop.ParquetFileReader;
import org.apache.parquet.hadoop.metadata.ParquetMetadata;
```
These 2 are satisfied by `org.apache.parquet:parquet-hadoop`. and then `import org.apache.parquet.Preconditions` (which should probably be replaced with `org.apache.iceberg.relocated.com.google.common.base.Preconditions`), which is being satisfied by `org.apache.parquet:parquet-common`. Both of these libs are part of `org.apache.parquet:parquet-avro` as can be seen below:

```

\--- org.apache.parquet:parquet-avro -> 1.12.0
     +--- org.apache.parquet:parquet-column:1.12.0
     |    +--- org.apache.parquet:parquet-common:1.12.0
     |    |    +--- org.apache.parquet:parquet-format-structures:1.12.0
     |    |    |    +--- org.slf4j:slf4j-api:1.7.22 -> 1.7.25
     |    |    |    \--- javax.annotation:javax.annotation-api:1.3.2
     |    |    +--- org.slf4j:slf4j-api:1.7.22 -> 1.7.25
     |    |    \--- org.apache.yetus:audience-annotations:0.12.0
     |    \--- org.apache.parquet:parquet-encoding:1.12.0
     |         \--- org.apache.parquet:parquet-common:1.12.0 (*)
     +--- org.apache.parquet:parquet-hadoop:1.12.0
     |    +--- org.apache.parquet:parquet-column:1.12.0 (*)
     |    +--- org.apache.parquet:parquet-format-structures:1.12.0 (*)
     |    +--- org.apache.parquet:parquet-jackson:1.12.0
     |    +--- org.xerial.snappy:snappy-java:1.1.8
     |    +--- commons-pool:commons-pool:1.6
     |    \--- com.github.luben:zstd-jni:1.4.9-1
     \--- org.apache.parquet:parquet-format-structures:1.12.0 (*)
```
Thus `org.apache.parquet:parquet-avro` carries everything we need for `:iceberg-data` in this case. Does that answer your question?


> Can you be more specific about what you're saying with implementation vs compile? When are transitive dependencies included and when are they not?

You could technically just replace the old `compile` configuration with the new `api` configuration, but according to  [Gradle docs](https://docs.gradle.org/current/userguide/java_library_plugin.html#java_library_plugin) it is better to use `implementation`, since compilation will be faster and other dependencies don't **leak** into your classpath during compilation. Below is the section from the Gradle docs that I'm referring to:

> Dependencies appearing in the api configurations will be transitively exposed to consumers of the library, and as such will appear on the compile classpath of consumers. Dependencies found in the implementation configuration will, on the other hand, not be exposed to consumers, and therefore not leak into the consumers' compile classpath. This comes with several benefits:

>* dependencies do not leak into the compile classpath of consumers anymore, so you will never accidentally depend on a transitive dependency

>* faster compilation thanks to reduced classpath size

less recompilations when implementation dependencies change: consumers would not need to be recompiled

cleaner publishing: when used in conjunction with the new maven-publish plugin, Java libraries produce POM files that distinguish exactly between what is required to compile against the library and what is required to use the library at runtime (in other words, don’t mix what is needed to compile the library itself and what is needed to compile against the library).

Comment (nastra) [reply build.gradle:253]: > Can you be more specific about what you're saying with implementation vs compile? When are transitive dependencies included and when are they not?

You could technically just replace the old `compile` configuration with the new `api` configuration, but according to  [Gradle docs](https://docs.gradle.org/current/userguide/java_library_plugin.html#java_library_plugin) it is better to use `implementation`, since compilation will be faster and other dependencies don't **leak** into your classpath during compilation. Below is the section from the Gradle docs that I'm referring to:

> Dependencies appearing in the api configurations will be transitively exposed to consumers of the library, and as such will appear on the compile classpath of consumers. Dependencies found in the implementation configuration will, on the other hand, not be exposed to consumers, and therefore not leak into the consumers' compile classpath. This comes with several benefits:

>* dependencies do not leak into the compile classpath of consumers anymore, so you will never accidentally depend on a transitive dependency

>* faster compilation thanks to reduced classpath size

>* less recompilations when implementation dependencies change: consumers would not need to be recompiled

>* cleaner publishing: when used in conjunction with the new maven-publish plugin, Java libraries produce POM files that distinguish exactly between what is required to compile against the library and what is required to use the library at runtime (in other words, don’t mix what is needed to compile the library itself and what is needed to compile against the library).

File: build.gradle (hunk 5 above)
Comment (RussellSpitzer) [reply build.gradle:178]: this is because Compile is now basically compileOnly

image

File: build.gradle [hunk 7]
```diff
@@ -570,15 +619,15 @@ if (jdkVersion == '8') {
...
-      compile project(':iceberg-orc')
+      implementation project(':iceberg-data')
+      implementation project(':iceberg-orc')
 
       testCompileOnly project(path: ':iceberg-data', configuration: 'testArtifacts')
       testCompileOnly project(path: ':iceberg-orc', configuration: 'testArtifacts')
     }
 
     shadowJar {
-      configurations = [project.configurations.compile, project.configurations.compileOnly, project.configurations.testCompileOnly]
```
Comment (RussellSpitzer) [inline build.gradle:581]: Why did this have a dependency on test compile deps ... that doesn't sound right.

Comment (nastra) [reply build.gradle:581]: don't know why it had this dependency but I agree that it wasn't correct

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 8]
```diff
@@ -306,6 +306,6 @@ private Schema lazyColumnProjection() {
   }
 
   private static String formatTimestampMillis(long millis) {
-    return DATE_FORMAT.format(LocalDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneId.systemDefault()));
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:309]: This is a slightly different behavior ... we used to use the System Default and now we are doing UTC?

File: core/src/main/java/org/apache/iceberg/FindFiles.java [hunk 9]
```diff
@@ -103,7 +103,7 @@ public Builder asOfTime(long timestampMillis) {
       // case, there is no valid snapshot to read.
       Preconditions.checkArgument(lastSnapshotId != null,
           "Cannot find a snapshot older than %s",
-          DATE_FORMAT.format(LocalDateTime.ofInstant(Instant.ofEpochMilli(timestampMillis), ZoneId.systemDefault())));
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/FindFiles.java:106]: Another slight change in behavior here ... I would probably save these for another PR unless we really need it here

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java (hunk 8 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:309]: yes, this was an error reported from the newer error-prone checks

> com.palantir.baseline-error-prone uses a newer error-prone version (and does additional checks), so a few changes were necessary to fix JavaTimeSystemDefaultTimeZone / ConsistentOverrides errors

File: core/src/main/java/org/apache/iceberg/FindFiles.java (hunk 9 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/FindFiles.java:106]: yes, this was an error reported from the newer error-prone checks

> com.palantir.baseline-error-prone uses a newer error-prone version (and does additional checks), so a few changes were necessary to fix JavaTimeSystemDefaultTimeZone / ConsistentOverrides errors

**Review by RussellSpitzer — APPROVED** [review]
Looks good to me, If it is possible I would try to separate out the build changes from the format changes if possible. Like first doing a bump of baseline if we can do that without the gradle change then doing the gradle change.

But I don't have strong feelings about that, and if all of our artifacts are the same before and after this Patch I think we are ok. I would recommend we double check the spark31 - spark tests to make sure their runtime class-paths are still correct.

**Comment by nastra** [comment]
> Looks good to me, If it is possible I would try to separate out the build changes from the format changes if possible. Like first doing a bump of baseline if we can do that without the gradle change then doing the gradle change.
> 
> But I don't have strong feelings about that, and if all of our artifacts are the same before and after this Patch I think we are ok. I would recommend we double check the spark31 - spark tests to make sure their runtime class-paths are still correct.

I tried to do that by creating 4 commits and keeping relevant things related to each other within a particular commit. I'm not sure doing the commits in reverse order is better here as I see those as logical steps: 1) Upgrade to Gradle 7 with related plugins that don't work otherwise 2) fix formatting or errorprone stuff separately due to newer versions in separate commits

File: core/src/main/java/org/apache/iceberg/FindFiles.java (hunk 9 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/FindFiles.java:106]: could we just skip that rule? We do for some other error prone rules. I'm fine with maybe making this change later just seems unrelated

**Comment by nastra** [comment]
@rdblue given that Iceberg 0.12.0 is out, I think now would be a good time to revisit this PR and get it merged.

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java [hunk 10]
```diff
@@ -90,11 +90,11 @@ public void testJsonConversion() throws Exception {
     long previousSnapshotId = System.currentTimeMillis() - new Random(1234).nextInt(3600);
     Snapshot previousSnapshot = new BaseSnapshot(
         ops.io(), previousSnapshotId, null, previousSnapshotId, null, null, null, ImmutableList.of(
-        new GenericManifestFile(localInput("file:/tmp/manfiest.1.avro"), SPEC_5.specId())));
+          new GenericManifestFile(localInput("file:/tmp/manfiest.1.avro"), SPEC_5.specId())));
```
Comment (aokolnychyi) [inline core/src/test/java/org/apache/iceberg/TestTableMetadata.java:93]: Are these required?

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java [hunk 11]
```diff
@@ -59,7 +59,7 @@ private CompatibilityHiveVectorUtils() {
    * Returns serialized mapwork instance from a job conf - ported from Hive source code LlapHiveUtils#findMapWork
    *
    * @param job JobConf instance
-   * @return
+   * @return A serialized {@link MapWork} based on the given job conf
```
Comment (aokolnychyi) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:62]: nit: We usually start with a lower case after `@return`.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java [hunk 12]
```diff
@@ -115,7 +115,7 @@ public String catalogName() {
   /**
    * The table properties string needed for the CREATE TABLE ... commands,
    * like "TBLPROPERTIES('iceberg.catalog'='mycatalog')
-   * @return
+   * @return The tables properties string, such as "TBLPROPERTIES('iceberg.catalog'='mycatalog')
```
Comment (aokolnychyi) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java:118]: nit: same here

Comment (aokolnychyi) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java:118]: I think quotes are not aligned too

**Review by aokolnychyi — COMMENTED** [review]
This PR looks mostly good to me. I had a question about the new scalastyle plugin license and a few minor nits.

I'll try to do another pass with fresh eyes tomorrow. The build file has a lot of changes around dependencies.

**Comment by aokolnychyi** [comment]
@kbendick @flyrain @karuppayya, could you help reviewing this one too if you have time?

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java (hunk 10 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestTableMetadata.java:93]: yes these are required

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java (hunk 11 above)
Comment (nastra) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:62]: I searched across the codebase and it seems that it's a bit inconsistent in terms of upper/lowercase. I generally think that something like this would be better if addressed by Checkstyle (but not sure if checkstyle supports such a check). Otherwise stuff like this will essentially pop up in lots of PR reviews

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java (hunk 12 above)
Comment (nastra) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java:118]: ah this is just a blind copy/paste from the javadoc description a few lines above without actually me checking whether it had missing pieces..will update

File: gradlew [hunk 13]
```diff
@@ -28,7 +44,7 @@ APP_NAME="Gradle"
 APP_BASE_NAME=`basename "$0"`
 
 # Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
-DEFAULT_JVM_OPTS=""
+DEFAULT_JVM_OPTS='"-Xmx64m" "-Xms64m"'
```
Comment (flyrain) [inline gradlew:47]: Is it necessary to set xmx here since we got this `org.gradle.jvmargs=-Xmx768m` in file gradle.properties? If we are going to use 768m as the xmx, we should just remove the 64m setting to avoid confusion.

File: gradlew.bat [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+set DIRNAME=%~dp0
+if "%DIRNAME%" == "" set DIRNAME=.
+set APP_BASE_NAME=%~n0
+set APP_HOME=%DIRNAME%
+
+@rem Resolve any "." and ".." in APP_HOME to make it shorter.
+for %%i in ("%APP_HOME%") do set APP_HOME=%%~fi
+
+@rem Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
+set DEFAULT_JVM_OPTS="-Xmx64m" "-Xms64m"
```
Comment (flyrain) [inline gradlew.bat:36]: Same here for `xmx`

File: gradlew (hunk 13 above)
Comment (nastra) [reply gradlew:47]: Those changes are generated automatically by Gradle when doing an upgrade

File: gradlew.bat (hunk 14 above)
Comment (nastra) [reply gradlew.bat:36]: This is also generated by Gradle itself when doing the upgrade

File: baseline.gradle [hunk 15]
```diff
@@ -47,7 +47,8 @@ subprojects {
 
   // So we apply Spotless manually to get a similar effect to baseline-format, but change the
   // import order.
-  pluginManager.withPlugin('com.diffplug.gradle.spotless') {
+
```
[the remark this answers is not in the document]
Comment (nastra) [reply baseline.gradle:50]: fixed

File: baseline.gradle [hunk 16]
```diff
@@ -82,4 +83,15 @@ subprojects {
...
+    scalastyle {
+      config = file("${rootDir}/project/scalastyle_config.xml")
+      inputEncoding = 'UTF-8'
+      outputEncoding = 'UTF-8'
+      failOnWarning = false
+      verbose = false
+      quiet = false
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (nastra) [reply baseline.gradle:97]: fixed

File: baseline.gradle [hunk 17]
```diff
@@ -36,7 +36,7 @@ subprojects {
     apply plugin: 'com.palantir.baseline-checkstyle'
     apply plugin: 'com.palantir.baseline-error-prone'
   }
-  apply plugin: 'com.palantir.baseline-scalastyle'
+  apply plugin: 'com.github.alisiikh.scalastyle'
```
[the remark this answers is not in the document]
Comment (nastra) [reply baseline.gradle:39]: I haven't found any license either, but `com.palantir.baseline-scalastyle` is essentially using `org.github.ngbinh.scalastyle` (see https://github.com/palantir/gradle-baseline/blob/develop/gradle-baseline-java/build.gradle#L15) and `com.github.alisiikh.scalastyle` is a fork of it (I guess because `org.github.ngbinh.scalastyle` was updated in 2018 the last time)

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java (hunk 12 above)
Comment (nastra) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java:118]: fixed

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java (hunk 11 above)
Comment (nastra) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:62]: fixed

File: gradle/wrapper/gradle-wrapper.properties [hunk 18]
```diff
@@ -19,6 +19,6 @@
 
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.1.1-bin.zip
```
Comment (snazy) [inline gradle/wrapper/gradle-wrapper.properties:22]: 7.2 has been released. Is it easy to use that version instead of 7.1.1?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java [hunk 19]
```diff
@@ -114,8 +114,8 @@ public String catalogName() {
 
   /**
    * The table properties string needed for the CREATE TABLE ... commands,
-   * like "TBLPROPERTIES('iceberg.catalog'='mycatalog')
-   * @return
+   * like TBLPROPERTIES('iceberg.catalog'='mycatalog')
+   * @return the tables properties string, such as TBLPROPERTIES('iceberg.catalog'='mycatalog')
```
Comment (snazy) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java:118]: nit: add `{@code ` around that TBLPROPERTIES

**Review by snazy — COMMENTED** [review]
LGTM

one nit and one ask to use Gradle 7.2 (released in the meantime), if easy to do

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java (hunk 19 above)
Comment (nastra) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestTables.java:118]: done

File: gradle/wrapper/gradle-wrapper.properties (hunk 18 above)
Comment (nastra) [reply gradle/wrapper/gradle-wrapper.properties:22]: good point, yes that should be easy to upgrade now.

**Comment by rdblue** [comment]
> Do we need to check in gradle-wrapper.jar? Wasn't it downloaded before?

We do not want to check in the gradle wrapper Jar because it conflicts with Apache release policy. We can't release a source tarball with the Jar in it and have to update the wrapper to download the Jar the first time it runs.

**Comment by nastra** [comment]
> > Do we need to check in gradle-wrapper.jar? Wasn't it downloaded before?
> 
> We do not want to check in the gradle wrapper Jar because it conflicts with Apache release policy. We can't release a source tarball with the Jar in it and have to update the wrapper to download the Jar the first time it runs.

Last time I checked locally, the wrapper wasn't downloaded automatically (not sure if they changed that). I just removed the wrapper jar and CI fails because the wrapper jar is missing
```
Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
Error: Process completed with exit code 1.
```

File: build.gradle [hunk 20]
```diff
@@ -201,44 +202,46 @@ project(':iceberg-bundled-guava') {
...
+    testImplementation project(path: ':iceberg-api', configuration: 'testArtifacts')
   }
 }
 
 project(':iceberg-data') {
   dependencies {
-    compile project(':iceberg-api')
-    compile project(':iceberg-core')
+    implementation project(path: ':iceberg-bundled-guava', configuration: 'shadow')
+    implementation project(':iceberg-api')
```
Comment (rdblue) [inline build.gradle:243]: I think this should be an `api` dependency because these classes are used in the API of `iceberg-data`. For example, `iceberg-api` includes the expressions, which are passed into `IcebergGenerics`.

File: build.gradle [hunk 21]
```diff
@@ -247,13 +250,30 @@ project(':iceberg-data') {
...
+    }
+
+    implementation("org.apache.parquet:parquet-avro") {
+      exclude group: 'org.apache.avro', module: 'avro'
+      // already shaded by Parquet
+      exclude group: 'it.unimi.dsi'
+      exclude group: 'org.codehaus.jackson'
+    }
+
+    compileOnly "org.apache.avro:avro"
```
Comment (rdblue) [inline build.gradle:268]: Why is Avro `compileOnly` when both `parquet-avro` and `orc-core` are `implementation`?

File: build.gradle [hunk 22]
```diff
@@ -1166,10 +1281,11 @@ project(':iceberg-spark3-runtime') {
   }
 
   dependencies {
-    compile project(':iceberg-spark3')
-    compile project(':iceberg-spark3-extensions')
-    compile project(':iceberg-aws')
-    compile(project(':iceberg-nessie')) {
+    implementation project(':iceberg-api')
```
Comment (rdblue) [inline build.gradle:1284]: I think this should be an `api` dependency here and most other places.

File: build.gradle [hunk 23]
```diff
@@ -1219,6 +1336,15 @@ project(':iceberg-spark3-runtime') {
...
     relocate 'org.antlr.v4.runtime', 'org.apache.iceberg.shaded.org.antlr.v4.runtime'
 
+    dependencies {
+      exclude 'org/antlr/runtime/**/*'
+      exclude 'org/abego/treelayout/**/*'
+      exclude 'com/ibm/icu/**/*'
+      exclude 'javax/json/**/*'
+      exclude 'org/glassfish/**/*'
+      exclude 'org/stringtemplate/**/*'
+    }
```
Comment (rdblue) [inline build.gradle:1346]: It is suspicious to me that this is now necessary. I think it is because we are shading `runtimeClasspath` now rather than `compile`. I think that change is mostly correct because `implementation` dependencies are no longer transitive unless you look at the runtime classpath. But some extra dependencies must be leaking in from the other changes.

I think the right way to exclude these is to add them to the `implementation` excludes above, not to exclude specific packages.

File: build.gradle (hunk 21 above)
Comment (nastra) [reply build.gradle:268]: it seems that `org.apache.avro:avro` was mostly used  as a `compileOnly` dependency in other places in this file, therefore I thought having it as `compileOnly` here would be good here as well

File: build.gradle (hunk 22 above)
Comment (nastra) [reply build.gradle:1284]: sgtm

File: build.gradle (hunk 20 above)
Comment (nastra) [reply build.gradle:243]: sgtm

File: build.gradle (hunk 23 above)
Comment (nastra) [reply build.gradle:1346]: oh you're totally right. I moved stuff to the exclude list of the `implementation` configuration and we're getting the same results

File: build.gradle [hunk 24]
```diff
@@ -498,24 +543,26 @@ project(':iceberg-hive-metastore') {
...
-    compile project(':iceberg-core')
-    compile project(':iceberg-data')
-    compile project(':iceberg-hive-metastore')
-    compile project(':iceberg-orc')
-    compile project(':iceberg-parquet')
+    implementation project(path: ':iceberg-bundled-guava', configuration: 'shadow')
+    api project(':iceberg-api')
+    implementation project(':iceberg-common')
+    implementation project(':iceberg-core')
+    implementation project(':iceberg-data')
```
Comment (rdblue) [inline build.gradle:562]: This should also be an API dependency because the `InputFormat` produces Iceberg classes in iceberg-data.

Comment (nastra) [reply build.gradle:562]: good point, updated this and `iceberg-spark`/`iceberg-flink` to an `api` dependency, since those were also using iceberg-data class in their API

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Merged. Thanks for working through all of the issues to get this done, @nastra!

File: build.gradle [hunk 25]
```diff
@@ -247,13 +250,30 @@ project(':iceberg-data') {
       exclude group: 'org.slf4j', module: 'slf4j-log4j12'
     }
 
-    testCompile("org.apache.hadoop:hadoop-client") {
+    implementation("org.apache.orc:orc-core::nohive") {
+      exclude group: 'org.apache.hadoop'
+      exclude group: 'commons-lang'
+      // These artifacts are shaded and included in the orc-core fat jar
+      exclude group: 'com.google.protobuf', module: 'protobuf-java'
```
Comment (shivjha30) [inline build.gradle:257]: Is it intended for the protobuf-java class coming from ORC, to be part of the runtime jar ?
For example: org/apache/iceberg/shaded/org/apache/orc/protobuf/CodedInputStream$ArrayDecoder.class
